### DISCUSSION TITLE
Hardhat artifacts and viem

### DISCUSSION THREAD

**Discussion by josemarinas** [NomicFoundation/hardhat#5066] [Troubleshooting] [answered]
Hello,

We are using Hardhat for compiling our contracts, we need to use the resulting artifacts with viem but we dont get the typesafety because they are not exported `as const` so viem does not detect them. Is there any way of solving this issue or whats the bes way of proceeding?

Thank you very much in advance

@fvictorio

**Comment by schaable** [comment]
Hi @josemarinas, are you using the `hardhat-viem` plugin? Can you describe what you're trying to do?

**Reply by josemarinas** [reply]
Hey @schaable 

Thank you for answering. Right now we are using `hardhat-ethers`, we want to export the artifacts in a way `viem`users can use the ABIs directly. The idea is to create an npm package with the generated artifacts and that viem users can just use the ABI and have typesafety.

**Reply by schaable** [reply]
`hardhat-viem` provides type generation for your contracts upon compilation. It automatically generates a type file for your contract under the `artifacts` directory, which includes the ABI. For instance, for the `Lock` contract, it generates something like this:

![image](https://github.com/NomicFoundation/hardhat/assets/1165424/54ec8908-ed50-4881-8e40-24d139cc04ae)

Is this what you're looking for?

**Reply by josemarinas** [reply]
Yep,  it seems like it however, I haven't been able to set it up properly.

I added the `hardhat-viem` and `viem` libraries

```bash
yarn add --dev hardhat-viem viem
```
Then i added this line to my `hardhat.config.ts`
```ts
// hardhat.config.ts
import '@nomicfoundation/hardhat-viem';
```

But the artifacts generated after building the contracts does not seem to change, however viem is available in the `hre` object. Is there something that I'm missing?

**Reply by schaable** [reply]
What do you mean by saying the artifacts don't seem to change? Typically, artifacts (including types) are regenerated each time you execute `hardhat compile`, provided there have been modifications to your contracts. Is that not happening for you? If that's the case, could you provide the steps your following so we can try to replicate?

**Reply by josemarinas** [reply]
I managed to generate the artifacts but now I didnt manage to use them with viem. My final objective is to crteate an npm package that holds this artifacts. I tried multiple stuff but they dont seem to be usable in TS outside of the Hardhat project.

I would like to use them like this

```ts
import * as artifacts from '..artifacts';
import hre from 'hardhat';
import {getContract} from 'viem';

async function main() {
  const contract = getContract({
    client,
    address: '0x1234567890123456789012345678901234567890',
    abi: artifacts.contract.abi // selecting the correct abi
  });

  const contract.write.function(params1, param2 ...)
}

main()
  .then(() => process.exit())
  .catch(error => {
    console.error(error);
    process.exit(1);
  });
```

**Reply by schaable** [reply]
That's not the correct way of using an artifact. Here´s an example:

```typescript
import hre from 'hardhat';
// ...
const artifact = await hre.artifacts.readArtifact("MyContract");
console.log(artifact.abi)
```

> they dont seem to be usable in TS outside of the Hardhat project.

If you want to read the artifacts generated by hardhat from outside the hardhat project, you could import them directly:

```typescript
import myContractArtifact from 'some/path/hardhat-project/artifacts/contracts/MyContract.sol/MyContract.json';
console.log(myContractArtifact.abi)
```

**Comment by mdesousa** [comment]
i was able to generate `abi.ts` by adding the subtask below to the hardhat config. it exports the abi as a constant, so it can be used as shown in [viem type inference](https://viem.sh/docs/typescript#type-inference). hope it helps!

```
type ContractMap = Record<string, { abi: object }>;

subtask(TASK_COMPILE_SOLIDITY_EMIT_ARTIFACTS).setAction(
  async (args, env, next) => {
    const output = await next();
    const { artifacts } = env.config.paths;
    const promises = Object.entries(args.output.contracts).map(
      async ([sourceName, contract]) => {
        const file = join(artifacts, sourceName, 'abi.ts');
        const { abi } = Object.values(contract as ContractMap)[0];
        const data = `export const abi = ${inspect(abi, false, null)} as const;`;
        await writeFile(file, data);
      },
    );
    await Promise.all(promises);
    return output;
  },
);
```

**Comment by 0xbuidler** [comment]
Hey @josemarinas , I'm facing the exact same issue. Did you figure it out ?

**Reply by josemarinas** [reply, marked as answer]
Yes, in the end I used [wagmi-cli](https://wagmi.sh/cli/getting-started) with this `wagmi.config.ts`, you'd need to update the paths to your own. Then you can import the generated abi with `viem`

```
import {defineConfig} from '@wagmi/cli';
import {hardhat} from '@wagmi/cli/plugins';

export default defineConfig({
  // using abi instead of generated to avoid being ignored by git
  out: `generated/abis.ts`,
  plugins: [
    hardhat({
      artifacts: `../contracts/artifacts/src`,
      exclude: ['**/test'],
      project: '../contracts',
    }),
  ],
});
```

**Reply by 0xbuidler** [reply]
Yeah, I ended up doing the same. Thanks anyway buddy 👌
